src/python/pants/engine/build_files.py

```diff
@@ -209,41 +206,28 @@ def _hydrate(item_type, spec_path, **kwargs):
   return item
 
 
-class _MappedSpecs(datatype([
-    ('address_families', tuple), # List[AddressFamily]
-    ('specs', Specs),
-])):
-  """Wraps up logic to resolve specs into individual target addresses, or error out."""
-
-  def __new__(cls, address_families, specs):
-    return super(_MappedSpecs, cls).__new__(cls, tuple(address_families), specs)
-
-  @memoized_property
-  def _exclude_compiled_regexps(self):
-    return [re.compile(pattern) for pattern in set(self.specs.exclude_patterns or [])]
-
-  def _excluded_by_pattern(self, address):
-    return any(p.search(address.spec) is not None for p in self._exclude_compiled_regexps)
-
-  @memoized_property
-  def _target_tag_matches(self):
-    def filter_for_tag(tag):
-      return lambda t: tag in [str(t_tag) for t_tag in t.kwargs().get("tags", [])]
-    return wrap_filters(create_filters(self.specs.tags, filter_for_tag))
-
-  def _matches_target_address_pair(self, address, target):
-    return self._target_tag_matches(target) and not self._excluded_by_pattern(address)
+@rule(BuildFileAddresses, [Select(AddressMapper), Select(Specs)])
+def addresses_from_address_families(address_mapper, specs):
+  """Given an AddressMapper and list of Specs, return matching BuildFileAddresses.
 
-  @memoized_property
-  def _address_family_by_directory(self):
-    return {af.namespace: af for af in self.address_families}
+  :raises: :class:`ResolveError` if:
+     - there were no matching AddressFamilies, or
+     - the Spec matches no addresses for SingleAddresses.
+  :raises: :class:`AddressLookupError` if no targets are matched for non-SingleAddress specs.
+  """
+  # Capture a Snapshot covering all paths for these Specs, then group by directory.
+  snapshot = yield Get(Snapshot, PathGlobs, _spec_to_globs(address_mapper, specs))
+  dirnames = {dirname(f.stat.path) for f in snapshot.files}
+  address_families = yield [Get(AddressFamily, Dir(d)) for d in dirnames]
+  address_family_by_directory = {af.namespace: af for af in address_families}
 
-  def _matching_addresses_for_spec(self, spec):
+  matched_addresses = OrderedSet()
```

Comment:
If this has specific relevance to the perf numbers described in the OP compared to the memoized property part, would leave a comment here in addition to this change.